### ISSUE TITLE
Support for additional `Content-Type`s

### ISSUE THREAD

**Issue by jlalmes** [trpc/trpc#1937]
If we go down this route what we're really talking about is straying away from JSON as the defacto mode of transport. If we do this we should also have support for some sort of binary format when doing `POST`s that could take the compression further and let us have something else than `content-type: application/json`.

Furthermore, I think the server should be able to receive different types of request types and be informed on what to do. I have thought about this a bit before in the context of transformers, so outlining below.

## Some ideas for communicating transformers/serializer

> Rough ideas below, I think best by writing code, but don't see any of this as a final design.

### `contentType`-param

Goals:

- Allow for multiple content-types
- Allow for deserializing from a `Response` with a `body`
- Allow for serializing into a `Fetch.body`
- Allow for serializing into a string for query params

**When creating a server:**

```ts
import { initTRPC } from '@trpc/server';
import { jsonContentType, createContentType } from '@trpc/server/content-type';

// JSON crush could be a community plugin that looks something like this:
import jsoncrush from 'jsoncrush';
const crush = createContentType({
  key: 'crush',
  contentType: 'application/json',
  // a smarter content type would use something that could be streamed, i.e ReadableStream from `Response`
  // used for `POST`:
  fromResponse: async (res) => jsoncrush.uncrush(await res.text()),
  fromString: (str) => jsoncrush.uncrush(str),
  toBody: data => jsoncrush.crush(data), // this could allow for the types that the `Fetch API` allows - ArrayBuffer, string, [...]
  // used for `GET`
  toString: data => jsoncrush.data(data),
})

const trpc = initTRPC({
  contentTypes: [
    jsonContentType,
    crush,
  }
});
```

**HTTP level**

```
// inform server that we are sending `crush` content
GET http://localhost?content=crush&input=[...]
GET http://localhost?content=json&input=[...]

POST http://localhost?content=crush
content-type: "application/json"
// [..]
```

### Similarly, `transformer`-param could be used

```ts
import { initTRPC } from '@trpc/server';
import { defaultTransform } from '@trpc/server/transform';
// JSON crush could be a community plugin
import { crush } from 'trpc-json-crush';

const trpc = initTRPC({
  transforms: [
      defaultTransform, // our "no-op" trnasform
      { key: "superjson", value: superjson },
   ],
});
```

### The client story

Maybe only needs to allow for one content type.

```ts
import { crush } from "~/shared/contentType"

const client = createTRPCClient({
  links: [ /* ... */ ],
  contentType: crush,
});

```

_Originally posted by @KATT in https://github.com/trpc/trpc/discussions/1936#discussioncomment-2818345_

**Comment by ecwyne** [comment]
This would be great! I was just looking to use https://www.npmjs.com/package/notepack.io with tRPC today and came up against json being the only possible encoding

**Comment by sachinraja** [comment]
I did an RFC in #2387. Addressing the comments in https://github.com/trpc/trpc/issues/2387#issue-1330907320

> I don't think "superjsonContentType" is a good one -- if we send a different content-type header on that we'll lose stuff like auto-formatting in the Network tab

I'm a little confused here, we wouldn't send a different content type header for something that is still JSON. Each content type could configure the value for that header.

> I think a link should be able to decide the content type for a specific procedure call. I.e. a splitLink can be used on a mutation like uploadProfilePhoto to define that it should be passed as binary

I agree, that's a good point.

> How do we deal with "mixing" content types,
For inputs, it's easy to split stuff with links
Let's say a query profile.photo returns a binary, we need to make sure that's not called in a batchLink that expects JSON in the response

Do we need to make sure of that? Seems unlikely this will actually error unexpectedly. And I don't see how that wouldn't be caught in development. We could include the content type key in the response body but that would require us to change the structure greatly.

> I think the "ending link" should be able to contain the logic of understanding different content types
httpBatchLink for JSON will probably be a distinctly different link than httpBatchLink for a binary types

Interesting idea, I would think we could extract all this logic to the content type object instead of having it in a link.

**Comment by lncramer** [comment]
+1 for this - I have a requirement to integrate with a webhook that expects a content type of `text/xml` in the response, and this is blocking me from being able to release.

Is there some other way I can currently achieve this? I'm using https://github.com/jlalmes/trpc-openapi to define the HTTP path for the webhook, but AFAICT there is no way for the associated tRPC procedure to return anything other than JSON (https://github.com/jlalmes/trpc-openapi/issues/58#issuecomment-1180631200).

**Comment by Azarattum** [comment]
I’m looking for a way to send binary data over WebSockets via tRPC subscriptions. Is this the issue that could allow that kind of functionality? So far only HTTP had been discussed here. Let me know if I should create a separate issue for binary over WebSockets.

**Comment by KATT** [comment]
> I’m looking for a way to send binary data over WebSockets via tRPC subscriptions. Is this the issue that could allow that kind of functionality? 

Ideally, we'd support this both for WSS and HTTP, but it's okay if we solve them one at a time, feel free to have a stab and write in the contributors channel on Discord if you want guidance!

**Comment by boredland** [comment]
Hi there! I would really love to return a buffer from a trpc route, so I could reuse all the middleware and things. 

This works just fine:

```typescript
ctx.res.setHeader('Content-Type', 'application/x-protobuf');
ctx.res.end(buffer);
```

Sadly 'end' terminates the response, so I get `error - Error [ERR_HTTP_HEADERS_SENT]: Cannot set headers after they are sent to the client`. Any idea how to work around that?

**Comment by harveylee** [comment]
I'm looking for a way to return a buffer also, thought about the same thing @boredland tried - but on top of the response termination error, I think [resolveHTTPResponse](https://github.com/trpc/trpc/blob/8d6dc2ddfba947d42167b7f9dd0016faf77c423e/packages/server/src/http/resolveHTTPResponse.ts#L61) will also transform and serialize the response to JSON unnecessarily on the server.

I'm trying to write an endpoint that returns output in the [Arrow IPC format](https://arrow.apache.org/docs/format/Columnar.html#serialization-and-interprocess-communication-ipc) so it's pretty wasteful marshaling it back and forth to JSON.

**Comment by AlexRMU** [comment]
It may be necessary to specify the types for serializers (transformers) more clearly.
For example, `defaultTransformer` is
```ts
type DataTransformer = {
    serialize(object: any): any;
    deserialize(object: any): any;
};
```
The default data type is JSON:
```ts
type TypeJSON =
  | string
  | number
  | boolean
  | null
  | { [x: string]: TypeJSON; }
  | Array<TypeJSON>
```
Both input and output are not checked in any way. And if, for example, you send a Set, the error will not be shown, although it will exist.
And what will happen when you can use a bunch of different serializers with a bunch of supported data types?

**Comment by boredland** [comment]
isn't the transformer already a tad too late down the drain ? afaik the assumption for transformers is to serialize sth. into a string of a JSON property. At least for my use-case (sending a buffer of the wire) that doesn't help too much.

**Comment by aprilmintacpineda** [comment]
This will be a good addition to TRPC, right now, I integrated TRPC with Fastify and have to use a custom fastify POST endpoint.

**Comment by EricDotSmith** [comment]
@KATT Are there any updates on this, specifically doing file uploads like with (png, pdf) using tRPC?

**Comment by ivi3** [comment]
Is there any plan to support content-type in TRPC v11?

**Comment by waltershewmake** [comment]
It seems like they are planning to add this based on the v.11 migration guide: https://trpc.io/docs/migrate-from-v10-to-v11#rawinput-in-middleware-is-now-a-getrawinput-breaking

**Comment by Nick-Lucas** [comment]
> It seems like they are planning to add this based on the v.11 migration guide: https://trpc.io/docs/migrate-from-v10-to-v11#rawinput-in-middleware-is-now-a-getrawinput-breaking

Yes we had a really good crack at this last year, and basically found we were unhappy with how eagerly tRPC loads inputs in V10 - basically could leave you a bit vulnerable to DDOS attacks without adding complexity in tRPC that the experimental-multipartform APIs we introduced needed. So the breaking change in V11 is to materialise inputs more lazily, which then allows us to deal in streams and other non-text inputs without materialising them for procedures that aren't expecting them!

**Comment by QimatLuo** [comment]
If I use v11 (@trpc/server@next), how do I customize my headers?
I cannot find documents teach me how to do.

**Comment by ci0n** [comment]
I need to upload zip or csv files for nodejs processing. These files are usually above 10mb.

**Comment by terion-name** [comment]
JSON is very readable but very expensive, especially on big payloads. Transport layer doesn't have to be readable and I was very surprised that trpc doesn't support transport customization. It's an obvious and needed feature. Using websocket transport with something like msgpack or capnproto encoder – binary and without parse-ahead – is a desired way for rpc

**Comment by gregg-cbs** [comment]
I found this repo that claims they upload form data with trpc but no matter what I do i cant get this to work:
https://github.com/trpc/examples-next-formdata/tree/main/src

If anyone has made any progress please comment here. Otherwise do i need to setup rest endpoints for file upload?

**Comment by okalil** [comment]
@gregg-cbs The approach of that repo appearently works for me. The key is using splitLink to change the data transformer:
```tsx
export const client = api.createClient({
  links: [
    loggerLink({
      enabled: (op) =>
        process.env.NODE_ENV === "development" ||
        (op.direction === "down" && op.result instanceof Error),
    }),
    splitLink({
      condition: (op) => isNonJsonSerializable(op.input),
      true: httpLink({
        url: getBaseUrl() + "/api/trpc",
        transformer: {
          serialize: (data) => data as FormData,
          deserialize: SuperJSON.deserialize,
        },
      }),
      false: unstable_httpBatchStreamLink({
        transformer: SuperJSON,
        url: getBaseUrl() + "/api/trpc",
      }),
    }),
  ],
});
```

Procedure:
```tsx
uploadFile: publicProcedure.input(z.instanceof(FormData)).mutation(...) // or use zod-form-data
```

Form:
```tsx
  const { mutate } = api.uploadFile.useMutation()
  // ... some code
      <form
        encType="multipart/form-data"
        action={(formData) => mutate(formData)}
      >
      // ... form fields
```

**Comment by Azarattum** [comment]
Would it be possible to send pure binary data through WebSockets with this new approach?

**Comment by busybox11** [comment]
From my understanding this does not support Fastify yet? Tried fiddling around with the code and looking at related PRs and changes, it looks like the Fastify tRPC plugin doesn't register `multipart/form-data` as a valid `Content-Type` header, leading to HTTP 415 Error `FST_ERR_CTP_INVALID_MEDIA_TYPE`.

For now it looks like unless you're using NextJS (which would make it trivial to implement it yourself anyway), the current way to go is to setup custom HTTP endpoints on your router, outside of tRPC.

Might be wrong, feel free to correct me if I am - I would love to be wrong :)

**Comment by busybox11** [comment]
So I slept on it, and got back to it today.
After LOTS of troubleshooting in the dark, I found myself stuck because of Zod, which led me to believe something was wrong with either tRPC or Fastify, as I didn't have any comprehensive stack trace anywhere and only got `failed to parse body as FormData` when using `attachFieldsToBody: "keyValues"` on the FormData Fastify plugin, among other errors on other situations.

Turns out removing Zod validation from a procedure does give you the data in the `input` variable with `keyValues` fields, which I have been able to verify by writing a tRPC middleware that hooks into the request body and logs it to the console.

Can't figure out how to make Zod work, but hey - that's progress! I was indeed wrong yesterday :)

**Comment by gregg-cbs** [comment]
I made my own trpc like client that i export from my backend and use on my front end. It was not easy but we can upload files and do typed requests. We also export some of our entity types so we have access to those on our client too.

I think the hardest part of it was getting typescript to work seamlessly, we are still solving some issues there but overall its a good experience that we desperately needed.

If Trpc could cover all cases i dont think we would have made our own thing but we were worried about needing features and being limited.

**Comment by irian-codes** [comment]
Alright I've found a solution to send FormData that I think it's the one I like for its modularity and conciseness. Which is using a shared custom transformer between the client and the client TRPC and server TRPC. I followed [the docs](https://trpc.io/docs/server/data-transformers#datatransformer-interface) on how to create it.

Hope you find it useful! 😊

**Shared transformer**
```ts
export const transformer: TRPCCombinedDataTransformer = {
  input: {
    serialize: (obj) => {
      if (isNonJsonSerializable(obj)) {
        return obj;
      } else {
        return superjson.serialize(obj);
      }
    },
    deserialize: (obj) => {
      if (isNonJsonSerializable(obj)) {
        return obj;
      } else {
        // eslint-disable-next-line
        return superjson.deserialize(obj);
      }
    },
  },
  output: superjson,
};
```

**Client TRPC, in my case, Next.js**
```ts
export const api = createTRPCNext<AppRouter>({
  config() {
    return {
      /**
       * Links used to determine request flow from client to server.
       *
       * @see https://trpc.io/docs/links
       */
      links: [
        loggerLink({
          enabled: (opts) =>
            process.env.NODE_ENV === 'development' ||
            (opts.direction === 'down' && opts.result instanceof Error),
        }),
        httpBatchLink({
          url: `${getBaseUrl()}/api/trpc`,
          transformer,
        }),
      ],
    };
  },
  /**
   * Whether tRPC should await queries when server rendering pages.
   *
   * @see https://trpc.io/docs/nextjs#ssr-boolean-default-false
   */
  ssr: false,
  transformer,
});
```

**Server TRPC**
```ts
const t = initTRPC.context<typeof createInnerTRPCContext>().create({
  transformer,
  errorFormatter({shape, error}) {
    return {
      ...shape,
      data: {
        ...shape.data,
        zodError:
          error.cause instanceof ZodError ? error.cause.flatten() : null,
      },
    };
  },
});
```

**Comment by manazoid** [comment]
@gregg-cbs Hello, can you share the solution?

**Comment by szagi3891** [comment]
@manazoid - in the client you need to plug decoding of this superjson format

```
import { superJsonCustom } from '@reactive/utils';

        this.client = createTRPCProxyClient<AppRouter>({
            transformer: superJsonCustom,
            links: [
            ....
```

**Comment by manazoid** [comment]
> ```
> import { superJsonCustom } from '@reactive/utils';
> ```

Where can I get this `@reactive/utils`

I get an error
```
error: GET https://registry.npmjs.org/@reactive%2futils - 404
```

**Comment by szagi3891** [comment]
My symbol "superJsonCustom" :

```
import superJsonCustom from 'superjson';

const arrayBufferToString = (buffer: ArrayBuffer): string => {
    const uint8Array = new Uint8Array(buffer);
    const hexString = [];
    for (const char of uint8Array) {
        const hex = char.toString(16).padStart(2, '0');
        hexString.push(hex);
    }

    return hexString.join('');
};

const stringToArrayBuffer = (data: string): ArrayBuffer => {
    const length = data.length / 2;
    const uint8Array = new Uint8Array(length);
    for (let i = 0; i < length; i++) {
        const byte = parseInt(data.substr(i * 2, 2), 16);
        uint8Array[i] = byte;
    }
    return uint8Array.buffer;
};

//type ArrayBuffer
superJsonCustom.registerCustom<ArrayBuffer, string>(
    {
        isApplicable: (v): v is ArrayBuffer => v instanceof ArrayBuffer,
        serialize: (v) => arrayBufferToString(v),
        deserialize: (v) => stringToArrayBuffer(v),
    },
    'ArrayBuffer'
);

export {
    superJsonCustom
};

```

**Comment by csvn** [comment]
The original proposal here looks very interesting. I was hoping to try out [CBOR](https://jsr.io/@std/cbor) on both client/server. Which is similar to Superjson, but creates a binary format instead (`application/cbor`). Unfortunately, this does not work with the [`transformer` property](https://trpc.io/docs/server/data-transformers#datatransformer-interface). Even though `DataTransformer ` has `serialize(object: any): any;`, `any` is more like JSON or `string` I guess, not `Uint8Array`.

It's great that it's possible to use [`File`, `Blob` and `Uint8Array`](https://trpc.io/docs/server/validators#file-and-other-binary-type-inputs) though. But this seems targeted to sending file/blob to server. Not being able to encode a large data payload as e.g. [MessagePack](https://msgpack.org/index.html) or CBOR.

**Comment by arxkdev** [comment]
Yeah I cannot get this solution working at all. I've tried everything above.

This is all I get:
```ts
❌ tRPC failed on <no-path>: config.transformer.output.serialize is not a function
 ⨯ [TypeError: config.transformer.output.serialize is not a function]
 ⨯ [TypeError: config.transformer.output.serialize is not a function]
```

**Comment by asimaranov** [comment]
> We do have the ability to support other content types these days, and overrides could be exposed on links and adapters fairly easily, so any PRs are very welcome there

Awesome! It would be extremely helpful if you could share your thoughts of how this can be implemented. Is it enough to implement custom transformer by extending DataTransformer? Maybe there are some references to see that could help? Thanks

**Comment by muhammedaksam** [comment]
This will save so many lives.

**Comment by kschniedergers** [comment]
^

**Comment by Maxim-Mazurok** [comment]
First time using tRPC, I was under impression it would use binary communication protocol like gRPC/protobuf, and I'm surprised that it doesn't... Seems like a fairly obvious feature since it controls all the serialization/deserialization, could be a good performance boost... My only concern going in was that debugging would be harder, since I won't be able to just open devtools and see responses as they would be binary (could be solved with browser ext?)
